controller/volume_rebuilding_controller.go

```diff
@@ -191,13 +318,122 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			va, err = vbc.syncLHVolumeAttachementForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
 
-	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol)
+		if err != nil {
+			return err
+		}
+		if isVolumeInRebuilding {
+			return nil
+		}
+	}
 
+	if isOfflineRebuildEnabled && vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		warnMsg := fmt.Sprintf("Volume %v is faulted, skip rebuilding", volName)
+		vbc.logger.Info(warnMsg)
+		vbc.eventRecorder.Event(vol, corev1.EventTypeWarning, constant.EventReasonCanceledOfflineRebuilding, warnMsg)
+	}
+
+	if !isOfflineRebuildEnabled && util.IsHigherPriorityVATicketExisting(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		warnMsg := fmt.Sprintf("Canceling volume %v offline rebuilding", volName)
+		vbc.logger.Info(warnMsg)
+		vbc.eventRecorder.Event(vol, corev1.EventTypeWarning, constant.EventReasonCanceledOfflineRebuilding, warnMsg)
+	}
+
+	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
 	return nil
 }
 
+func (vbc *VolumeRebuildingController) isVolumeOfflineRebuildEnabled(offlineRebuilding longhorn.VolumeOfflineRebuilding) (bool, error) {
+	offlineReplicaRebuildingValue, err := vbc.ds.GetSettingAsBool(types.SettingNameOfflineReplicaRebuilding)
+	if err != nil {
+		return false, err
+	}
+	return (offlineRebuilding == longhorn.VolumeOfflineRebuildingEnabled) || (offlineRebuilding == longhorn.VolumeOfflineRebuildingIgnored && offlineReplicaRebuildingValue), nil
+}
+
+func (vbc *VolumeRebuildingController) syncLHVolumeAttachementForOfflineRebuild(vol *longhorn.Volume, va *longhorn.VolumeAttachment, attachmentID string) (*longhorn.VolumeAttachment, error) {
+	replicas, err := vbc.ds.ListVolumeReplicasRO(vol.Name)
+	if err != nil {
+		return va, err
+	}
+	areAllReplicasHealthy, err := vbc.isVolumeReplicasHealthy(vol.Spec.NumberOfReplicas, replicas)
+	if err != nil {
+		return va, err
+	}
+	if !areAllReplicasHealthy {
+		createOrUpdateAttachmentTicket(va, attachmentID, vol.Status.OwnerID, longhorn.TrueValue, longhorn.AttacherTypeVolumeRebuildingController)
+	}
+	return va, nil
+}
+
+func (vbc *VolumeRebuildingController) isVolumeReplicasHealthy(numberOfReplicas int, replicas map[string]*longhorn.Replica) (bool, error) {
+	healthyCount := 0
+	for _, replica := range replicas {
+		node, err := vbc.ds.GetNodeRO(replica.Spec.NodeID)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return false, err
+		}
+		isNodeReady := false
+		if node.Status.Conditions != nil {
+			nodeReadyCondition := types.GetCondition(node.Status.Conditions, longhorn.NodeConditionTypeReady)
+			isNodeReady = nodeReadyCondition.Status == longhorn.ConditionStatusTrue
+		}
```

Comment:
I think we don't need to check the node condition and can focus solely on the volume replicas.
If there's an issue with the node, it will be reflected through the replicas status.

Reply:
>If there's an issue with the node, it will be reflected through the replicas status.

After some tests, the replica status just switched the `OwnerID` to another worker node when a worker node shut down was detected.

Reply:
We switched to monitoring the replica status instead of the node status.